pkg/kubectl/cmd/delete.go

```diff
@@ -95,20 +96,59 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		return err
 	}
 
+	// By default use a reaper to delete all related resources.
+	if cmdutil.GetFlagBool(cmd, "cascade") {
+		return ReapResult(r, f, out, cmd)
+	}
+	return DeleteResult(r, out)
+}
+
+func ReapResult(r *resource.Result, f *cmdutil.Factory, out io.Writer, cmd *cobra.Command) error {
 	found := 0
-	err = r.IgnoreErrors(errors.IsNotFound).Visit(func(r *resource.Info) error {
+	err := r.IgnoreErrors(errors.IsNotFound).Visit(func(info *resource.Info) error {
 		found++
-		if err := resource.NewHelper(r.Client, r.Mapping).Delete(r.Namespace, r.Name); err != nil {
+		reaper, err := f.Reaper(info.Mapping)
+		if err != nil {
+			// If the error is "not found" and the user didn't explicitly set "cascade" just delete.
+			if kubectl.IsNoSuchReaperError(err) && cmd.Flags().Lookup("cascade") != nil && !cmd.Flags().Lookup("cascade").Changed {
```

Comment:
Repeating since it may have been lost in outdated diff

Checking the flag for nil is kind of an ugly hack... How about this: change ReapResult to take a boolean `defaultToDelete` instead of passing the Command object. If `defaultToDelete` is true, do a standard delete on resources for which there is no reaper available. That way you can move the check for an explicit `--cascade` into RunDelete, and we're not checking for nonexistent flags as a proxy for "which toplevel command is calling me".